pr_agent/servers/azuredevops_server_webhook.py

```diff
@@ -0,0 +1,135 @@
+# This file contains the code for the Azure DevOps Server webhook server.
+# The server listens for incoming webhooks from Azure DevOps Server and forwards them to the PR Agent.
+# ADO webhook documentation: https://learn.microsoft.com/en-us/azure/devops/service-hooks/services/webhooks?view=azure-devops
+
+import json
+import os
+import re
+import secrets
+import uvicorn
+from fastapi import APIRouter, Depends, FastAPI, HTTPException
+from fastapi.security import HTTPBasic, HTTPBasicCredentials
+from fastapi.encoders import jsonable_encoder
+from starlette import status
+from starlette.background import BackgroundTasks
+from starlette.middleware import Middleware
+from starlette.requests import Request
+from starlette.responses import JSONResponse
+from starlette_context.middleware import RawContextMiddleware
+
+from pr_agent.agent.pr_agent import PRAgent, command2class
+from pr_agent.algo.utils import update_settings_from_args
+from pr_agent.config_loader import get_settings
+from pr_agent.git_providers.utils import apply_repo_settings
+from pr_agent.log import get_logger
+from fastapi import Request, Depends
+from fastapi.security import HTTPBasic, HTTPBasicCredentials
+from pr_agent.log import get_logger
+
+security = HTTPBasic()
+router = APIRouter()
+available_commands_rgx = re.compile(r"^\/(" + "|".join(command2class.keys()) + r")\s*")
+azure_devops_server = get_settings().get("azure_devops_server")
+WEBHOOK_USERNAME = azure_devops_server.get("webhook_username")
+WEBHOOK_PASSWORD = azure_devops_server.get("webhook_password")
+
+def handle_request(
+    background_tasks: BackgroundTasks, url: str, body: str, log_context: dict
+):
+    log_context["action"] = body
+    log_context["api_url"] = url
+    with get_logger().contextualize(**log_context):
+        background_tasks.add_task(PRAgent().handle_request, url, body)
+
+
+# currently only basic auth is supported with azure webhooks
+# for this reason, https must be enabled to ensure the credentials are not sent in clear text
+def authorize(credentials: HTTPBasicCredentials = Depends(security)):
+        is_user_ok = secrets.compare_digest(credentials.username, WEBHOOK_USERNAME)
+        is_pass_ok = secrets.compare_digest(credentials.password, WEBHOOK_PASSWORD)
+        if not (is_user_ok and is_pass_ok):
+            raise HTTPException(
+                status_code=status.HTTP_401_UNAUTHORIZED,
+                detail='Incorrect username or password.',
+                headers={'WWW-Authenticate': 'Basic'},
+            )
```

Comment:
/ask make sure to use https be enabled to ensure the credentials are not sent in clear text